deploy/buildtools/buildfarmhostdispatcher.py

```diff
@@ -0,0 +1,302 @@
+import logging
+import sys
+
+from awstools.awstools import *
+
+# typing imports
+from typing import Any, Dict, Optional, List, TYPE_CHECKING
+from mypy_boto3_ec2.service_resource import Instance
+# needed to avoid type-hint circular dependencies
+# TODO: Solved in 3.7.+ by "from __future__ import annotations" (see https://stackoverflow.com/questions/33837918/type-hints-solve-circular-dependency)
+#       and normal "import <module> as ..." syntax (see https://www.reddit.com/r/Python/comments/cug90e/how_to_not_create_circular_dependencies_when/)
+if TYPE_CHECKING:
+    from buildtools.buildconfig import BuildConfig
+else:
+    BuildConfig = object
+
+rootLogger = logging.getLogger()
+
+class BuildFarmHostDispatcher(object):
+    """Abstract class to manage how to handle a single build farm host (request, wait, release, etc).
+
+    Attributes:
+        NAME: Human-readable name (used as the "build-farm-type" in the YAML).
+        build_config: Build config associated with the dispatcher.
+        args: Set of args/options associated with the dispatcher.
+        is_local: Is the dispatcher the same as the manager host?
+        override_remote_build_dir: Override name of build dir on build host (default to $HOME).
+    """
+    NAME: str = ""
```

Comment:
This is a @timsnyder-siv question, but yeah. That seems perfectly sensible, and seems to be what `SO` suggests. 

There's also this: https://docs.python.org/3/library/abc.html

Reply:
I would stay away from getters and setters unless there is high overhead to getting or setting the member.  So, not in this case since there aren't any expensive side-effects from looking up the stringy name of the class. 

I think you should definitely look into inheriting from the stdlib's `abc.ABC` instead of rolling your own implementation of an AbstractBaseClass.  See https://docs.python.org/3/library/abc.html that David pointed to.   Thanks for the pointer David, I had forgotten that this is one of the Python3 things I like.  It should also simplify 

 You should define it as an abstract class-property, only implement the property getter and mark it with the `@abstractmethod` decorator. I think that `@abstractmethod` will move the implementation protections to compile-time errors, so you don't _need_ your abstract methods to `raise NotImplemented` but it would prevent someone in a derived class from doing something like `return super().NAME` and expecting it to give a legit value (since apparently abstract methods in Python can have an implementation but it can only be called via an explicit reference through `super()`).
 
 So, to be concrete, do this:
```suggestion
    @property
    @abstractmethod
    @classmethod
    def NAME() -> str:
        raise NotImplementedError
```

And you need your base class to derive from `abc.ABC` instead of `class`.

Oh, and `NAME` in all caps is the correct casing for a class-wide constant per https://google.github.io/styleguide/pyguide.html#3164-guidelines-derived-from-guidos-recommendations

Reply:
Looking at this again, you would either need to add a single `cls` parameter to `NAME(cls)` or use `@staticmethod` instead.

Reply:
Ok, I implemented this as the static class version.